File: src/test/java/ru/jerael/booktracker/backend/web/mapper/GenreWebMapperTest.java
package ru.jerael.booktracker.backend.web.mapper;

import org.junit.jupiter.api.Test;
import ru.jerael.booktracker.backend.web.dto.genre.GenreResponse;
import ru.jerael.booktracker.backend.domain.model.Genre;
import java.util.Set;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertEquals;

class GenreWebMapperTest {
    private final GenreWebMapper genreWebMapper = new GenreWebMapper();

    @Test
    void toResponse() {
        Genre genre = new Genre(1, "adventure");

        GenreResponse genreResponse = genreWebMapper.toResponse(genre);

        assertEquals(1, genreResponse.id());
        assertEquals("adventure", genreResponse.name());
    }

    @Test
    void toResponses() {
        Genre genre1 = new Genre(1, "adventure");
        Genre genre2 = new Genre(2, "action");
        Set<Genre> genres = Set.of(genre1, genre2);

        Set<GenreResponse> genreResponses = genreWebMapper.toResponses(genres);

        assertEquals(2, genreResponses.size());
        assertThat(genreResponses).extracting(GenreResponse::id).containsExactlyInAnyOrder(1, 2);
        assertThat(genreResponses).extracting(GenreResponse::name).containsExactlyInAnyOrder("adventure", "action");
    }
}

File: src/test/kotlin/ru/jerael/booktracker/backend/web/mapper/GenreWebMapperTest.kt
package ru.jerael.booktracker.backend.web.mapper

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import ru.jerael.booktracker.backend.factory.genre.GenreDomainFactory

class GenreWebMapperTest {
    private val mapper = GenreWebMapper()
    
    @Test
    fun `toResponse should map Genre to GenreResponse`() {
        val genre = GenreDomainFactory.createGenre()
        
        val response = mapper.toResponse(genre)
        
        with(response) {
            assertEquals(genre.id, id)
            assertEquals(genre.name, name)
        }
    }
    
    @Test
    fun `toResponses should map Genres to GenreResponses`() {
        val genre1 = GenreDomainFactory.createGenre(1, "Genre 1")
        val genre2 = GenreDomainFactory.createGenre(2, "Genre 2")
        val genres = setOf(genre1, genre2)
        
        val responses = mapper.toResponses(genres)
        
        val expectedResponses = genres.map { mapper.toResponse(it) }.toSet()
        assertEquals(expectedResponses, responses)
    }
}